llvm/include/llvm/IR/NVVMIntrinsicFlags.h

```diff
@@ -0,0 +1,40 @@
+//===--- NVVMIntrinsicFlags.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// This file contains the definitions of the enumerations and flags
+/// associated with NVVM Intrinsics.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_NVVMINTRINSICFLAGS_H
+#define LLVM_SUPPORT_NVVMINTRINSICFLAGS_H
+
+#include <stdint.h>
+
+namespace llvm {
+namespace nvvm {
+
+enum class CpAsyncBulkTensorLoadMode {
+  TILE = 0,
+  IM2COL = 1,
+};
+
+typedef union {
+  int V;
+  struct {
+    unsigned CacheHint : 1;
+    unsigned MultiCast : 1;
+    unsigned LoadMode : 3; // CpAsyncBulkTensorLoadMode
+    unsigned reserved : 27;
+  } U;
+} CpAsyncBulkTensorFlags;
```

Comment:
Type-punning union fields is generally not a good idea in C++.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c183-dont-use-a-union-for-type-punning

Can you elaborate on why we need to do that? AFAICT, we seem to use it to pass just one i32 argument to the instruction node, as opposed to specifying cache/multicast/loadmode separately.

Is there a reason we can not pass each parameter separately?

Reply:
The value of this union is used as the 'flag' operand.
This is the first i32 operand of intrinsic which must be a compile-time constant.
This encodes the modifiers and load modes which then the backend
peeks at and lowers to the corresponding instructions. This is not
passed to PTX but used only from IR to ISel.

The usage of the 'flag' enables us to have one intrinsic (per-dim) here
instead of the 8 intrinsics (2(ch)x 2(mc) x 2(mode)) required otherwise.

For example, in the G2S case:
* The optional cache_hint is always passed to the intrinsic as an i64 operand.
  If the cache_hint bit is set, the i64 operand is used and the intrinsic
  gets lowered to the _CH variant in the backend. Since this is optional,
  when the cache_hint bit is not set, the corresponding i64 operand is ignored
  (can be an undef) during codegen.
* Similarly, depending on the multicast and cache_hint bits, the
  intrinsic lowers to one of the four variants:
  no_MC_CH(default), only _MC, only _CH, both _MC_CH.
* The same approach is extended for the load-modes too.

It is possible to accommodate additional modifiers in future archs,
by extending the flag-bits instead of doubling the number of
intrinsics for each modifier.

I hope this clarifies the behavior of the flags. I would be
happy to discuss more details offline. (I have pinged you
on discord)

Reply:
OK. It's a reasonable approach, though multiple intrinsics would not be unreasonable, either, especially if we could automatically generate them via tablegen. PTX instructions do tend to be ... rather non-unifirm in their variants, so I can see the attractiveness of just passing a single opaque value and handling its bits as needed.
My concern was mostly about the mechanics of conversion between the opaque integer and the bits. Considering that the value never leaves the boundaries of the compilation, we do not need to worry about the bitfield order in the integer, but I's still prefer to do it without type-punning.

AFAICT, before c++20 the only valid way to convert between the union fields is via memcpy. In c++20 one would use `std::bitcast` to do the job, but we're not there yet, as LLVM build with c++17 at the moment.

Alternatively, you could use a class/struct with the fields, and define explicit methods to convert to/from an integer, and avoid the type-punning issues altogether.